Use dl_esm_inf infrastructure

arporter opened the pull request:
This is for #8 and removes the GOcean infrastructure code from this project and imports it from dl_esm_inf instead. I've also added dl_timer as a second git submodule as that reduces the need for user configuration (i.e. they don't need to tell us where dl_timer is and/or build it).

arporter commented:
@rupertford if you're looking to add manual OmpSs and I want to add/update the OpenACC implementation, please could you review this ticket as the re-structuring it contains will make it easier... Thanks!

rupertford commented on ocean/nemo/nemolite2d/manual_versions/psykal_ocl/fortran/Makefile:
```diff
@@ -0,0 +1,42 @@
+# OpenCL libraries and any flags from environment variables. e.g. to
+# use AMD's dispatch library:
+# 
+# export F90=gfortran
```
Are these commented out compile options required?

arporter replied:
No, they're just to show a user what they need to set before running make.

rupertford commented on ocean/nemo/nemolite2d/manual_versions/psykal_acc/Makefile.include:
```diff
@@ -0,0 +1 @@
+../manual_acc/Makefile.include
```
could we point the actual path to the manual_acc version, rather than having a sym link?

rupertford commented on ocean/nemo/nemolite2d/manual_versions/psykal_ocl/fortran/Makefile:
```diff
@@ -0,0 +1,42 @@
+inf_lib:
+	make -C ${INF_DIR}
+
+nemolite2d.exe: inf_lib ${COMMON_MODULES}
```
I think this line should be above timer_lib, otherwise the Makefile will only compile timer_lib when you type make.

rupertford commented on ocean/nemo/nemolite2d/manual_versions/psykal_ocl/fortran/gocean2d_io_mod.f90:
```diff
@@ -0,0 +1 @@
+../../../psykal/gocean2d_io_mod.f90
```
I think the Makefile should have a set of shared files from a different directory, rather than sym linking them.

arporter replied:
Yes. I've re-worked things so that we now have a `nemolite2d/common/` directory that contains these files.

rupertford commented on ocean/nemo/nemolite2d/manual_versions/psykal_omp/Makefile:
```diff
@@ -0,0 +1,83 @@
+inf_lib:
+	make -C ${INF_DIR}
+
+all: $(EXECS)
```
this should be above timer_lib, otherwise it won't be compiled with make.

arporter replied:
Doh.

arporter commented:
I've created a separate `kernels` directory and put both the Fortran and OpenCL kernels in it. I've then removed all the kernels from the various nemolite implementations and updated the Makefiles to simply create links to the necessary Fortran source files. I'm happy with the way I've done this for `nemolite2d/psykal` but less so for the others - I still need to update them.

arporter commented:
You can have git submodules point to any version you want so yes, we could have them point to releases. (We haven't yet made a release of dl_esm_inf though.)
I think I've addressed all of the comments.
Ready for 2nd review...

arporter commented:
Although I've added READMEs for the OpenCL versions, they contain TBC's because I'm still working on the OpenCL implementation(s) in #11 and need to straighten things out.

arporter commented:
All benchmarks now pick-up the compiler settings etc. from environment variables.
This is documented in the README.md file in each directory. Each of these now have `Compiling` and `Running` sections.
The README for the 'original' NEMOLite now explains why infrastructure is needed.
The README for nemolite2d/psykal explains that it uses code generation. The various manual versions that were in that directory are now in manual_versions/psykal_serial.
make clean does.
I've added a top-level Makefile that builds all the CPU benchmarks. This might facilitate the later incorporation of some noddy CI.
The Makefiles all build the various libraries first (if required).
If there's some sort of record for the number of README's written in one day then I think I should get it!
Back to @rupertford.

rupertford commented:
Much better. Yes you get the Hartree README creation award for 2018.

Could you please sort out the Makefile in ocean/nemo/nemolite2d/psykal as "make" does not build the executables as it says it does in the README.

arporter commented:
Argh! Done now.

(Do I get a certificate?)

rupertford commented:
[ARP_Certificate.pdf](https://github.com/stfc/PSycloneBench/files/2084990/ARP_Certificate.pdf)
